### PR TITLE
Create a database interface

### LINKED ISSUE #39: Create database facade and implementation
**Describe what to enhance (and how if already theorized)**
We want a facade that our app can interface with that can abstract the implementation of the database we're using. This facade could call a wrapper class for the specific database being used. To avoid scope creep, this implementation can simply include basic CRUD for each of our models, with more complex queries being handled in a follow up story.

### PR DESCRIPTION
**Pull request by rmcreyes**
# :ship: Pull Request

## :question: Purpose

**Give a brief description of your changes:**
This change creates the objects to be used for database accesses. This includes a wrapper object for handling MySQL operations as well as an adapter that can be used to connect the underlying database query implementation (like the aforementioned MySQL implementation) to the rest of the application. The queries supported right now are simple CRUD operations for each of our current models. Some models needed to be edited to be given editable properties in order to support this. Some custom errors were also created to make error handling cleaner later on.

## :tickets: Ticket(s)

Closes #39

### CODE REVIEW

File: src/db/drivers/mysql.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+   * @returns promise that resolves with true if a user was deleted and false
+   *          if that user cannot be found, and rejects with an error message
+   */
+  public deleteUser(id: number): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      this.sequelizeUserModel.destroy({ where: { id: id } })
+        .then(rowsDeleted => {
+          // assuming it is impossible for rowsDeleted to be greater than 1, because if
+          // we're autoincrementing IDs we should be deleting at most one user
+          if (rowsDeleted < 1) {
```
Comment (andreamah) [inline src/db/drivers/mysql.ts]: Shouldn't rowsDeleted be either 1 or 0? Why are you testing for any integer < 1? If a -1 is possible for error in the destroy function, shouldn't that be treated differently?

File: src/db/drivers/mysql.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        host: host,
+        dialect: "mysql"
+      }
+    );
+
+    // define our user model for sequelize
+    this.sequelizeUserModel = this.sequelize.define("User", {
+      id: {
+        primaryKey: true,
+        type: DataTypes.INTEGER.UNSIGNED,
```
Comment (andreamah) [inline src/db/drivers/mysql.ts:89]: Should allowNull be false for the id values?

File: src/db/drivers/mysql.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  }
+
+  /**
+   * Adds a user to the users table.
+   *
+   * @param username username of the user to create
+   * @returns promise containing newly created user if resolved,
+   *          and an error message if rejected
+   */
+  public createUser(username: string): Promise<User> {
```
Comment (andreamah) [inline src/db/drivers/mysql.ts:224]: Perhaps add "TODO: add password field with proper hashing for security" to remind us that this method is incomplete

**Review by andreamah — COMMENTED** [review]
Amazing progress to get started!

**Comment by rmcreyes** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/db/errors.ts  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rmcreyes/keener-backend/pullRequest?prid=5536169&bid=18148722)

File: src/db/drivers/mysql.ts (hunk 1 above)
Comment (rmcreyes) [reply src/db/drivers/mysql.ts]: Used a different implementation so as to avoid this issue.

File: src/db/drivers/mysql.ts (hunk 2 above)
Comment (rmcreyes) [reply src/db/drivers/mysql.ts:89]: Done

**Comment by rmcreyes** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/db/errors.ts  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rmcreyes/keener-backend/pullRequest?prid=5536169&bid=18148722)

**Comment by rmcreyes** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/db/errors.ts  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rmcreyes/keener-backend/pullRequest?prid=5536169&bid=18148722)

**Pull request merged**